### ISSUE TITLE
Map user stories

### ISSUE THREAD

**Issue by simonrnss** [AI-SDC/AI-SDC#141]
Map user stories for imaginable use cases to enable us to know what the researcher will provide, what will be passed between code stages, etc.

Placeholder -- to be refined.

**Comment by simonrnss** [comment]
Actors: researchers vs output checkers
supported model types vs safe models vs non-programmatic

**Comment by jim-smith** [comment]
Main document is on teams in WP5/files folder:

**Comment by jim-smith** [comment]
- User creates an object “mydata” of type aisdc.attacks.dataset.Data and provides a separate code file that does the translation between the data in the format provided and the data in the format to be input to the machine any model. 
- User creates a model “mymodel” from the safeXClassifier class and calls mymodel.fit(). 
- User calls mymodel.preliminary_check() to make sure their hyper-parameters are within the TRE risk appetite for algorithm X. 
- User calls mymodel.run_attack(mydata) for different attack types and iterates over different hyper-parameters until they have an accurate model, and they interpret attack results as safe. 
- User calls myModel.request_release() with parameters modelsavefile.sav and  again passing the mydata object (without it request_release does not run attacks). 
  - LIRA, worst_case, and attribute_inference attacks are run automatically,  
  - results are stored ready for the TRE output checkers to look at. 
  - System also saves the results of mymodel.posthoc_check() for poor practice, model edits etc. 
- TRE checker has everything they need to make a decision with no further processing.

**Comment by jim-smith** [comment]
2: Next case
- User provides Data object and code, uses safeXClassifier() but does not pass data object to request_release() or save processed form of data. 
- safeXClassifer report checks for class disclosure and TRE risk appetite for algorithm X. 
- TRE output checker has to manually recreate processed data using code provided. 
- TRE output checker is unable to run any attacks UNLESS they also know exactly which rows from the dataset were used for training and testing. 
- So dataset object needs to store those specific details OR use fixed values for seed (e.g. to sklearn.train_test_split() ) and be extremely transparent about how stratification was done). 
- If TRE has enough info to recreate train/test processed data, then they can
    -    run attacks from script. 
    - Then the post-processing script 
    - Then make a judgement.

**Comment by jim-smith** [comment]
## 3: User provides dataset object but does not use safeXClassifier
- In this case we don’t currently have any checking for TRE-approved hyper-parameters or for class disclosure. 
  - But if it is a type where we have a safemodel version, we could create functionality to load it and then check hyper-parameters using existing code 
  - This raises the issue of whether safeModelClassifiers should have a load() option ?? – I;s currently commented out 
  - Could also provide method for checking for k-anonymity (and possible pure nodes) where appropriate by refactoring safemodels. 
- TREs need to manually configure and start scripts to do LIRA, Worst_Case and Attribute_Inference attacks 
   - NB this assumes their classifier outputs probabilities.

**Comment by jim-smith** [comment]
## 4: User does not use safeXClassifier, or provide dataset object
### but does provide description of pre-processing, 
### and provides output probabilities for the train and test set they have used (and true classes?) 
- We cannot assume that the TRE has the capability to get the right bits of pre-processing code from their source code.  
- Do we insist on this (would be needed for ‘outside world’)? what if this is commercially sensitive?  
- TRE can in theory run LIRA and worst-case but not attribute inference attacks. 
  - There is a risk that they have misidentified the train/test splits to give us ones which make the classifier look less disclosive 
  - But this probably falls outside our remit? 
- Recommend reject???   
-We could automate generalisation (as lower bound) and  worst case attacks if they give output probabilities 
   – so we need to specify format 
- TRE would need actual copies of processed data to run LIRA 

** THIS would be the version that let people use R **

**Comment by jim-smith** [comment]
## 5:  User creates differentially private algorithm (not via our code) and provides sufficient details to create data object. 
- How do we know what the actual epsilon value is? 
- If it is a keras model we can reload and query it if they have stored the training object as part of the model save (we need epochs, dataset size, L2 norm clip, noise values).    
  -  But then their stored model probably has disclosive values in anyway … 
  -   So would have to delete before release. 
  - And anyway, are keras models safe against attacks that change ‘trainable’ to true for different layers and then do repeated queries viz, attacks of federated learning. 
- If it is non keras then do, we take it on trust?? 
  - Probably yes that comes under safe researcher?? 

- TRE can recreate processed training and test sets and run attacks. 
- Does the actual epsilon value matter if we are doing that? 
   - Yes probably, because it is the sort of thing a TRE may well set as a policy.

**Comment by jim-smith** [comment]
## 6 Worst Case
- User makes R model for a tree-based classifier that we have not experimented with.  
- TREs get researcher to provide at minimum the processed train and test files. 

- From those we can’t run LIRA (because what would shadow models be?) 
-  but we can  worst-case from the command line or a script if their model outputs probabilities. 
-  And we can measure generalisation error. 
-  But not attribute inference. 
- We have no way of checking against class disclosure e.g. all training items in a specific subgroup ending in a ‘pure’ node. 

- Very hard to check and recommend release

**Comment by yolaj-nhs** [comment]
- [ ] Case 1
- [ ] Case 2
- [ ] Case 3
- [ ] Case 4
- [ ] Case 5
- [ ] Case 6

**Comment by yolaj-nhs** [comment]
[summary_user_story_3.txt](https://github.com/AI-SDC/AI-SDC/files/11947853/summary_user_story_3.txt)
[user_story_3.zip](https://github.com/AI-SDC/AI-SDC/files/11947857/user_story_3.zip)

**Comment by yolaj-nhs** [comment]
[summary_user_story_1.txt](https://github.com/AI-SDC/AI-SDC/files/11950265/summary_user_story_1.txt)



[user_story_1.zip](https://github.com/AI-SDC/AI-SDC/files/11950267/user_story_1.zip)

**Comment by jim-smith** [comment]
@Yola: these are great.
I've started separating out your code  for journey3 into two files:
 - an example of what a researcher might do
 - a single py file that TREs can run to generate the reports
 
Now I'm struggling to test them because my (apple silicon) macbook won't install the latest version of aisdc :(

**Comment by yolaj-nhs** [comment]
> @yola: these are great. I've started separating out your code for journey3 into two files:
> 
> * an example of what a researcher might do
> * a single py file that TREs can run to generate the reports
> 
> Now I'm struggling to test them because my (apple silicon) macbook won't install the latest version of aisdc :(

Oh no - that's unfortunate! Please feel free to send any updated files my way, and I'll test them on my end

**Comment by yolaj-nhs** [comment]
[summary_user_story_2_best_case.txt](https://github.com/AI-SDC/AI-SDC/files/11950588/summary_user_story_2_best_case.txt)
[user_story_2_best_case.zip](https://github.com/AI-SDC/AI-SDC/files/11950592/user_story_2_best_case.zip)
[user_story_2_worst_case.zip](https://github.com/AI-SDC/AI-SDC/files/11950593/user_story_2_worst_case.zip)

**Comment by jim-smith** [comment]
@yolaj-nhs can you confirm what branch I should be testing this in please?
I'm currently working in development.

Inside example_notebooks I created a folder 'user journeys' and subfolders inside for 'resources' then one for each journey
- I'm still working on journey 3.

Then I split your file into two as per the attached files (rename .txt to .py to run)
- you'll see in the researcher one where it is tweaks to pick up the 'right' version of acro for my system

[researcher_code.txt](https://github.com/AI-SDC/AI-SDC/files/11950927/researcher_code.txt)
[tre_generate_report.txt](https://github.com/AI-SDC/AI-SDC/files/11950928/tre_generate_report.txt)

Running the researcher one does what you woulds expect.
Running the TRE script  gives this error message right at the end:
%python tre_generate_report.py

Creating disclosure report for saved model using data splits provided ...

Reading model from training_artefacts//disclosive_random_forest.sav
Reading training/testing data from ./training_artefacts/
Now running attacks
...worst case
...Likelihood Ratio
INFO:_check_and_update_dataset:new ytrain has values and counts: (array([0, 1, 2, 3]), array([1297, 1247, 1256,   88]))
INFO:_check_and_update_dataset:new ytest has values and counts: (array([0, 1, 2, 3]), array([3023, 3019, 2788,  240]))
INFO:lr-scenario:Training shadow models
INFO:lr-scenario:Trained 0 models
INFO:lr-scenario:Trained 10 models
INFO:lr-scenario:Trained 20 models
INFO:lr-scenario:Trained 30 models
INFO:lr-scenario:Trained 40 models
INFO:lr-scenario:Trained 50 models
INFO:lr-scenario:Trained 60 models
INFO:lr-scenario:Trained 70 models
INFO:lr-scenario:Trained 80 models
INFO:lr-scenario:Trained 90 models
INFO:lr-scenario:Computing scores for train rows
INFO:lr-scenario:Computing scores for shadow rows
INFO:lr-scenario:Finished scenario
INFO:reporting:Starting report, report_name = None
attacks run ...compiling report
processing json files
Invalid command. Try --help to get more details
error message is 'GenerateTextReport' object has no attribute 'process_attack_target_json'

**Comment by yolaj-nhs** [comment]
> @yolaj-nhs can you confirm what branch I should be testing this in please? I'm currently working in development.
> 
> Inside example_notebooks I created a folder 'user journeys' and subfolders inside for 'resources' then one for each journey
> 
> * I'm still working on journey 3.
> 
> Then I split your file into two as per the attached files (rename .txt to .py to run)
> 
> * you'll see in the researcher one where it is tweaks to pick up the 'right' version of acro for my system
> 
> [researcher_code.txt](https://github.com/AI-SDC/AI-SDC/files/11950927/researcher_code.txt) [tre_generate_report.txt](https://github.com/AI-SDC/AI-SDC/files/11950928/tre_generate_report.txt)
> 
> Running the researcher one does what you woulds expect. Running the TRE script gives this error message right at the end: %python tre_generate_report.py
> 
> Creating disclosure report for saved model using data splits provided ...
> 
> Reading model from training_artefacts//disclosive_random_forest.sav Reading training/testing data from ./training_artefacts/ Now running attacks ...worst case ...Likelihood Ratio INFO:_check_and_update_dataset:new ytrain has values and counts: (array([0, 1, 2, 3]), array([1297, 1247, 1256, 88])) INFO:_check_and_update_dataset:new ytest has values and counts: (array([0, 1, 2, 3]), array([3023, 3019, 2788, 240])) INFO:lr-scenario:Training shadow models INFO:lr-scenario:Trained 0 models INFO:lr-scenario:Trained 10 models INFO:lr-scenario:Trained 20 models INFO:lr-scenario:Trained 30 models INFO:lr-scenario:Trained 40 models INFO:lr-scenario:Trained 50 models INFO:lr-scenario:Trained 60 models INFO:lr-scenario:Trained 70 models INFO:lr-scenario:Trained 80 models INFO:lr-scenario:Trained 90 models INFO:lr-scenario:Computing scores for train rows INFO:lr-scenario:Computing scores for shadow rows INFO:lr-scenario:Finished scenario INFO:reporting:Starting report, report_name = None attacks run ...compiling report processing json files Invalid command. Try --help to get more details error message is 'GenerateTextReport' object has no attribute 'process_attack_target_json'

Oh I'm so sorry - this is based on the refactor_summary_output branch, I should have been clearer! The user_stories_resources folder included above is also needed

**Comment by jim-smith** [comment]
@yolaj-nhs runs fine in that branch.   

One thing it does do is leave some files in the top-level directory(I.e. where my two  code files are)
- lira_config.json and loc_roc.png,
- it would be nice if they didn't;t get left there, shouldn't they be in the 'training artefacts folder? 

Also:
1.  we should change my scripts so they don't have hard-coded folder names
2. This user journey illustrates the problem that we end up with the model(to be released) in the same folder as the data (not to be released).

**Comment by jim-smith** [comment]
@yolaj-nhs we should perhaps have an issue for this.  For now here're updated version of of my scripts, where the tRE one doesn't assume any directory or file names.

so needs running (in this example) as:
python tre_generate_report.py --directory "requested_outputs" --model-file "disclosive_random_forest.sav"


but in general calling it with --help gives:
usage: tre_generate_report.py [-h] [--model-file MODEL_FILE] [--directory DIRECTORY]
                              [--trainX-file TRAINX_FILE] [--testX-file TESTX_FILE]
                              [--trainy-file TRAINY_FILE] [--testy_file TESTY_FILE]

Generate a risk report from a saved model and datasets

options:
  -h, --help            show this help message and exit
  --model-file MODEL_FILE
                        Filename for the saved ML model. Default = model.sav.
  --directory DIRECTORY
                        Name of directory holder saved model and data. Default = outputs.
  --trainX-file TRAINX_FILE
                        Filename for the saved training data. Default = trainX.txt.
  --testX-file TESTX_FILE
                        Filename for the saved test data. Default = testX.txt.
  --trainy-file TRAINY_FILE
                        Filename for the saved training data labels. Default = trainy.txt.
  --testy_file TESTY_FILE
                        Filename for the saved test labels. Default = testy.txt.

**Comment by yolaj-nhs** [comment]
## 7: User provides safemodel with no data
- User loads in data and pre-processes out with Target object
- User uses SafeDecisionTreeClassifier 
- User calls request_release() themselves, but does not pass data object to request_release() or save processed form of data.
- SafeDecisionTreeClassifier report checks for class disclosure and TRE risk appetite for algorithm X.
- User may send the dataset to TRE, but does not provide details of pre-processing, nor gives details about which samples were used for training/testing
- TRE has to rely on their own judgement and what the researcher has told them - AISDC in this case cannot provide any additional assistance
